src/mgmt/srv_system.c

```diff
@@ -0,0 +1,848 @@
+		goto out_lock;
+	}
+
+	rc = rdb_tx_commit(&tx);
+	if (rc != 0) {
+		D_ERROR("failed to commit map version %u: %d\n", map_version,
+			rc);
+		goto out_lock;
+	}
+
+	D_DEBUG(DB_TRACE, "rank %u joined in map version %u\n", rank,
+		map_version);
+	svc->ms_map_version = map_version;
+	if (in->ji_rank != -1)
+		svc->ms_rank_next = rank_next;
+	notify_map_distributord(svc);
+	out->jo_rank = rank;
+	out->jo_flags = SERVER_IN;
+
+out_lock:
+	ABT_rwlock_unlock(svc->ms_lock);
+	rdb_tx_end(&tx);
+out_svc:
+	mgmt_svc_put_leader(svc);
+out:
+	return rc;
+}
+
+static int
+map_update_bcast(crt_context_t ctx, struct mgmt_svc *svc, uint32_t map_version,
+		 int nservers, struct server_entry servers[])
+{
+	struct mgmt_tgt_map_update_in  *in;
+	struct mgmt_tgt_map_update_out *out;
+	crt_opcode_t			opc;
+	crt_rpc_t		       *rpc;
+	int				rc;
+
+	D_DEBUG(DB_MGMT, "enter: version=%u nservers=%d\n", map_version,
+		nservers);
+
+	opc = DAOS_RPC_OPCODE(MGMT_TGT_MAP_UPDATE, DAOS_MGMT_MODULE, 1);
+	rc = crt_corpc_req_create(ctx, NULL /* grp */,
+				  NULL /* excluded_ranks */, opc,
+				  NULL /* co_bulk_hdl */, NULL /* priv */,
+				  0 /* flags */,
+				  crt_tree_topo(CRT_TREE_KNOMIAL, 32), &rpc);
+	if (rc != 0) {
+		D_ERROR("failed to create system map update RPC: %d\n", rc);
+		goto out;
+	}
+	in = crt_req_get(rpc);
+	in->tm_servers.ca_count = nservers;
+	in->tm_servers.ca_arrays = servers;
+	in->tm_map_version = map_version;
+
+	rc = dss_rpc_send(rpc);
+	if (rc != 0)
+		goto out_rpc;
+
+	out = crt_reply_get(rpc);
+	if (out->tm_rc != 0)
+		rc = -DER_IO;
+
+out_rpc:
+	crt_req_decref(rpc);
+out:
+	D_DEBUG(DB_MGMT, "leave: version=%u nservers=%d: %d\n", map_version,
+		nservers, rc);
+	return rc;
+}
+
+static int
+distribute_map(crt_context_t ctx, struct mgmt_svc *svc)
+{
+	struct rdb_tx		tx;
+	struct enum_server_arg	arg = {};
+	int			rc;
+
+	rc = rdb_tx_begin(svc->ms_rsvc.s_db, svc->ms_rsvc.s_term, &tx);
+	if (rc != 0)
+		goto out;
+
+	ABT_rwlock_rdlock(svc->ms_lock);
+
+	rc = rdb_tx_iterate(&tx, &svc->ms_servers, false /* !backward */,
+			    enum_server_cb, &arg);
+	if (rc != 0)
+		goto out_lock;
+
+	rc = map_update_bcast(ctx, svc, svc->ms_map_version,
+			      arg.esa_servers_len, arg.esa_servers);
+
+	D_FREE(arg.esa_servers);
+out_lock:
+	ABT_rwlock_unlock(svc->ms_lock);
+	rdb_tx_end(&tx);
+out:
+	return rc;
+}
+
+static void
+map_distributord(void *arg)
+{
+	struct mgmt_svc	       *svc = arg;
+	struct dss_module_info *info = dss_get_module_info();
+
+	for (;;) {
+		bool	step_down;
+		int	rc;
+
+		ABT_mutex_lock(svc->ms_mutex);
+		for (;;) {
+			step_down = svc->ms_step_down;
+			if (step_down)
+				break;
+			if (svc->ms_distribute) {
+				svc->ms_distribute = false;
+				break;
+			}
+			ABT_cond_wait(svc->ms_distribute_cv, svc->ms_mutex);
+		}
+		ABT_mutex_unlock(svc->ms_mutex);
+		if (step_down)
+			break;
+		rc = distribute_map(info->dmi_ctx, svc);
+		if (rc != 0)
+			/* Try again. */
+			svc->ms_distribute = true;
+		dss_sleep(1000 /* ms */);
```

Comment:
I wonder should we put all ult schedule policy in one place, so we can have a better and controllable ult scheduler.

Reply:
Hmm, I'd like to sleep even if there no other ULT/tasklet needs to run.

Reply:
Oh, I do not question the sleep behavior. Just want to say whether this 1000 ms sleep schedule policy should be merged to dss_sched_run(), so we have a "single" place to schedule these ULT by their nature.

Reply:
We agree this is something the scheduler needs to be involved. I filed https://jira.hpdd.intel.com/browse/DAOS-505 for this.